File: cpp/35_数组中的逆序对.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    long cnt = 0;
    std::vector<int> tmp;
    
    // 外部接口函数
    int InversePairs(std::vector<int> nums) {
        int length = nums.size();
        if (length == 0) {
            return 0;
        }
        
        tmp = nums;
        mergeSort(nums, 0, length - 1);
        return int(cnt % 1000000007);
    }

    // 递归调用的函数
    // 注意使用引用，否则排序后无法替换原数组的值
    void mergeSort(std::vector<int>& nums, int l, int h) {
        if (l >= h) {
            // 当分组到只剩一个时，返回上一层，开始merge
            return;
        }
        int m = l + (h - l) / 2;
        mergeSort(nums, l, m);     // 左数组，统计，排序
        mergeSort(nums, m + 1, h); // 右数组，统计，排序
        merge(nums, l, m, h);      // 本次的前后数组，统计，排序
    }
    
    // 用来实现当前前后数组统计与合并的函数
    // 将两个有序数组合并为一个，同时统计逆序对
    // 注意使用引用
    void merge(std::vector<int>& nums, int l, int m, int h) {
        int i = l, j = m + 1, k = l;
        while (i <= m || j <= h) {
            if (i > m) { // 此时前数组已被遍历完，所有前数组小于后数组当前值
                tmp[k++] = nums[j++];
            } else if (j > h) { // 后数组被遍历完
                tmp[k++] = nums[i++];
            } else if (nums[i] <= nums[j]) { // 两个数组都没有遍历完，前 < 后
                tmp[k++] = nums[i++];
            } else { // 两个数组都没有遍历完，前 > 后，符合逆序数要求
                tmp[k++] = nums[j++];
                this->cnt += m - i + 1;
            }
        }
        
        // 将完成排序的部分拷贝到原数组中
        for (k = l; k <= h; k++) {
            nums[k] = tmp[k];
        }
    }
};

int main(int argc, char* argv[])
{
    Solution s;
    std::vector<int> nums = {1, 2, 3, 4, 5, 6, 7, 0};

    std::cout << s.InversePairs(nums) << std::endl;
    return 0;
}

File: python/35_数组中的逆序对.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

class Solution():
    def __init__(self):
        self.cnt = 0

    def InversePairs(self, nums):
        """
        注意，这种方式没有对外界的nums进行保护
        """
        if nums == []:
            return 0
        self.mergeSort(nums, 0, len(nums) - 1)
        return self.cnt % 1000000007
        
    def mergeSort(self, nums, l, h):
        if l >= h:
            return
        m = l + (h - l) // 2
        self.mergeSort(nums, l, m)
        self.mergeSort(nums, m + 1, h)
        self.merge(nums, l, m, h)

    def merge(self, nums, l, m, h):
        i = l
        j = m + 1
        k = l
        tmp = []
        while i <= m or j <= h:
            if i > m: # 前数组已经遍历完
                tmp.append(nums[j])
                j += 1
            elif j > h or nums[i] <= nums[j]: # 后数组已经遍历完，或则　前 < 后
                tmp.append(nums[i])
                i += 1
            else: # 前 > 后，符合逆序对的定义
                tmp.append(nums[j])
                j += 1
                self.cnt += m - i + 1
            k += 1

        # 拷贝数组
        nums[l:h+1] = tmp[:]


def main():
    s = Solution()
    nums = [1, 2, 3, 4, 5, 6, 7, 0]
    print(s.InversePairs(nums))


if __name__ == "__main__":
    main()